packages/upload-api/src/index/lib/sharded-dag-index.js

```diff
@@ -0,0 +1,109 @@
+import { CARReaderStream } from 'carstream'
+import * as dagCBOR from '@ipld/dag-cbor'
+import { ok, error, Schema, Failure, base58btc } from '@ucanto/server'
+import * as Digest from 'multiformats/hashes/digest'
+import * as API from './api.js'
+import { DigestMap } from './digest-map.js'
+
+export const ShardedDAGIndexSchema = Schema.variant({
+  'index/sharded/dag@0.1': Schema.struct({
+    /** DAG root. */
+    content: Schema.link(),
+    /** Shards the DAG can be found in. */
+    shards: Schema.array(Schema.link()),
+  }),
+})
+
+export const BlobIndexSchema = Schema.tuple([
+  /** multihash bytes. */
+  Schema.bytes(),
+  Schema.array(
+    /** multihash bytes, offset, length. */
+    Schema.tuple([Schema.bytes(), Schema.number(), Schema.number()])
+  ),
+])
+
+/**
+ * @param {ReadableStream<Uint8Array>} archive
+ * @returns {Promise<API.Result<API.ShardedDAGIndex, API.DecodeFailure|API.UnknownFormat>>}
+ */
+export const extract = async (archive) => {
```

Comment:
Other than CAR decoding, lot of the stuff here is provided by the schema facility e.g. `ShardedDAGIndexSchema.match` does quite a bit of work see here for some details https://github.com/web3-storage/ucanto/blob/main/packages/core/src/message.js

P.S. `view` function from link expects result from [`CAR.decode`](https://github.com/web3-storage/ucanto/blob/main/packages/core/src/car.js#L101) with `blocks` renamed to `store` (I know this was silly).

Reply:
ACK, I have split into `view` and `extract`.